common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -375,6 +357,44 @@ void initialize(String hostname, OpenSSLKey channelIdPrivateKey) throws IOExcept
         setTlsChannelId(channelIdPrivateKey);
     }
 
+    void configureServerCertificate() throws IOException {
+        verifyWithSniMatchers(getRequestedServerName());
+        if (isClient()) {
+            return;
+        }
+        X509KeyManager keyManager = parameters.getX509KeyManager();
+        if (keyManager != null) {
+            for (String keyType : getCipherKeyTypes()) {
+                try {
+                    setCertificate(aliasChooser.chooseServerAlias(keyManager, keyType));
+                } catch (CertificateEncodingException e) {
+                    throw new IOException(e);
+                }
+            }
+        }
+    }
+
+    private void verifyWithSniMatchers(String serverName) {
+        if (!AddressUtils.isValidSniHostname(serverName)) {
+            return;
+        }
+
+        if (Platform.allSniMatchersFail(parameters, serverName)) {
+            throw new IllegalArgumentException("SNI match failed: " + serverName);
```

Comment:
We shouldn't throw IllegalArgumentException here, the serverName is coming from the SNI header, so it's user-provided and could be anything.  This should probably throw SSLHandshakeException.